File: eyelet/input/graph.c
/*
 *  graph.c
 *
 *  gcc -std=c11 -O2 -Wall -Wextra graph.c -o graph
 *  ./path2n
 *
 *  Expected output:
 *      paris
 *      chartres
 *      lemans
 *      angers
 */

#include <stdio.h>
#include <stdbool.h>

/* ------------------------------------------------------  graph data -- */

/* 1.  City identifiers (order is arbitrary but must match the names[]) */
enum City {
    PARIS, ORLEANS, CHARTRES, AMIENS, BLOIS,
    BOURGES, TOURS, LEMANS, ANGERS, NANTES,
    N_CITIES                       /* <- automatically counts them     */
};

static const char *const names[N_CITIES] = {
    [PARIS]   = "paris",
    [ORLEANS] = "orleans",
    [CHARTRES]= "chartres",
    [AMIENS]  = "amiens",
    [BLOIS]   = "blois",
    [BOURGES] = "bourges",
    [TOURS]   = "tours",
    [LEMANS]  = "lemans",
    [ANGERS]  = "angers",
    [NANTES]  = "nantes"
};

/* 2.  Edge list — literally the oneway/2 facts                       */
struct Edge { enum City from, to; };

static const struct Edge edges[] = {
    { PARIS,    ORLEANS  },
    { PARIS,    CHARTRES },
    { PARIS,    AMIENS   },
    { ORLEANS,  BLOIS    },
    { ORLEANS,  BOURGES  },
    { BLOIS,    TOURS    },
    { CHARTRES, LEMANS   },
    { LEMANS,   ANGERS   },
    { LEMANS,   TOURS    },
    { ANGERS,   NANTES   }
};

static const int N_EDGES = sizeof edges / sizeof edges[0];

/* -------------------------------------------------  depth-first path -- */

static bool dfs(enum City src, enum City dst,
                bool visited[N_CITIES])
{
    if (src == dst) return true;
    visited[src] = true;

    for (int i = 0; i < N_EDGES; ++i)
        if (edges[i].from == src) {
            enum City v = edges[i].to;
            if (!visited[v] && dfs(v, dst, visited))
                return true;
        }
    return false;
}

static bool path(enum City a, enum City b)
{
    bool visited[N_CITIES] = { false };
    return dfs(a, b, visited);
}

/* -----------------------------------------------------------  query -- */

int main(void)
{
    for (enum City c = 0; c < N_CITIES; ++c)
        if (c != NANTES && path(c, NANTES))
            puts(names[c]);

    return 0;
}

File: eyelet/input/graph.py
#!/usr/bin/env python3
"""
Answer the Prolog-style query  ?- path(X, nantes).
For every solution X, print one shortest path X → … → nantes
as proof evidence.

No external libraries required – just the Python std-lib.
"""

from collections import defaultdict, deque

# ──────────────────────────────────────────────────────────────
# 1.  Ground “oneway” facts  (directed graph edges)
# ──────────────────────────────────────────────────────────────
EDGES: list[tuple[str, str]] = [
    ("paris",   "orleans"),
    ("paris",   "chartres"),
    ("paris",   "amiens"),
    ("orleans", "blois"),
    ("orleans", "bourges"),
    ("blois",   "tours"),
    ("chartres","lemans"),
    ("lemans",  "angers"),
    ("lemans",  "tours"),
    ("angers",  "nantes"),
]

GOAL = "nantes"

# Build forward and reverse adjacency lists once -----------------------------
succs: dict[str, set[str]] = defaultdict(set)   # u → {v, …}
preds: dict[str, set[str]] = defaultdict(set)   # v → {u, …}
for u, v in EDGES:
    succs[u].add(v)
    preds[v].add(u)

# ──────────────────────────────────────────────────────────────
# 2.  Reverse-BFS from the goal
#    parent[U] = V  means   U → … → V → … → GOAL   (shortest chain)
# ──────────────────────────────────────────────────────────────
def bfs_backwards(goal: str) -> dict[str, str]:
    parent: dict[str, str] = {}
    q: deque[str] = deque([goal])

    while q:
        v = q.popleft()
        for u in preds[v]:            # follow edge  u → v
            if u not in parent:       # not visited yet
                parent[u] = v
                q.append(u)
    return parent                     # keys = all sources that reach goal

# ──────────────────────────────────────────────────────────────
# 3.  Reconstruct one shortest path  src → … → goal
# ──────────────────────────────────────────────────────────────
def path_to_goal(src: str, goal: str, parent: dict[str, str]) -> list[str]:
    path = [src]
    while path[-1] != goal:
        path.append(parent[path[-1]])
    return path

# ──────────────────────────────────────────────────────────────
# 4.  Main
# ──────────────────────────────────────────────────────────────
if __name__ == "__main__":
    parent_map = bfs_backwards(GOAL)

    sources = sorted(parent_map.keys())        # all X with path(X, nantes)

    print("Solutions to  ?- path(X, nantes).   (and one proof each)\n")
    for src in sources:
        proof = path_to_goal(src, GOAL, parent_map)
        chain = " → ".join(proof)
        # left-align src in an 8-char field just for tidy columns
        print(f"X = {src:<8}   proof:  {chain}")

